core/domain/SuspicionDetection.py

```diff
@@ -0,0 +1,197 @@
+"""Sucpicion detection class which performs various detections from frame."""
+import numpy as np
+
+from core.classifiers import EventDetector
+from core.classifiers import Inception
+from core.classifiers import UnusualActivityDetector
+from core.classifiers import YOLOClassifier
+from core.platform.async import async
+import vgconf
+
+
+class SuspicionDetection(object):
+    async = async.Async()
+
+    def __init__(self):
+        self.yolo_buffer = []
+        self.inception_buffer = []
+        self.activity_detector_buffer = []
+        self.event_detector_buffer = []
+
+        self._yolo_buffer_lock = self.async.get_lock()
+        self._inception_buffer_lock = self.async.get_lock()
+        self._activity_detector_buffer_lock = self.async.get_lock()
+        self._event_detector_buffer_lock = self.async.get_lock()
+
+        self.is_inception_on = False
+        self.inception = None
+
+        self.is_event_detector_on = False
+        self.event_detector = None
+
+        self.is_activity_detector_on = False
+        self.activity_detector = None
+        self.activity_detector_length = vgconf.ACTIVITY_DETECTOR_LENGTH
+
+        self.is_yolo_on = False
+        self.yolo = None
+
+        self.yolo_sample_rate = vgconf.DEFAULT_YOLO_SAMPLE_RATE
+        self.inception_sample_rate = vgconf.DEFAULT_INCEPTION_SAMPLE_RATE
+
+        self.count = 0
+        self.sample_rate_lcm = self._get_sample_rate_lcm()
+
+    def _get_gcd(self, a, b):
+        while b > 0:
+            a, b = b, a % b
+        return a
+
+    def _get_lcm(self, a, b):
+        return (a * b) / self._get_gcd(a, b)
+
+    def _get_sample_rate_lcm(self):
+        return self._get_lcm(self.yolo_sample_rate, self.inception_sample_rate)
+
+    def _get_random_input(self, shape):
+        return np.random.uniform(size=shape)
+
+    def _get_event_detector(self):
+        self.event_detector = EventDetector.EventDetector()
+        random_input = self._get_random_input((2048,))
+        x = self.event_detector.predict(random_input)
+        self.is_event_detector_on = True
+
+    def _get_inception(self):
+        self.inception = Inception.Inception()
+        random_input = self._get_random_input((299, 299, 3))
+        x = self.inception.predict(random_input)
+        self.is_inception_on = True
+
+    def _get_unusual_activity_detector(self):
+        self.activity_detector = (
+            UnusualActivityDetector.UnusualActivityDetector())
+        random_input = self._get_random_input((4, 2048))
+        x = self.activity_detector.predict(random_input)
+        self.is_activity_detector_on = True
+
+    def _get_yolo_classifier(self):
+        self.yolo = YOLOClassifier.YOLO()
+        random_input = self._get_random_input((299, 299, 3))
+        x = self.yolo.predict(random_input)
+        self.is_yolo_on = True
+
+    def _remove_event_detector(self):
+        # How to do this ???
```

Comment:
This is still remaining. I don't have how to deallocate tensorflow / keras models and free up GPU memory.

Reply:
def limit_mem():
    K.get_session().close()
    cfg = K.tf.ConfigProto()
    cfg.gpu_options.allow_growth = True
    K.set_session(K.tf.Session(config=cfg))

try this. hope it will work.

Reply:
**READ THIS**
"http://forums.fast.ai/t/tip-clear-tensorflow-gpu-memory/1979/7"

Note that we do not release memory, since that can lead to even worse memory fragmentation.

What I think happens here: calling K.get_session() in the first line creates a session with default config, which uses all the memory. session.close() doesn’t release it, hence memory consumption stays the same as it was without calling limit_mem.

**Just putting this block of code in the beginning of the notebook will work:**

from keras import backend as K
cfg = K.tf.ConfigProto()
cfg.gpu_options.allow_growth = True
K.set_session(K.tf.Session(config=cfg))